extension/observer/ecsobserver/sd.go

```diff
@@ -17,41 +17,104 @@ package ecsobserver
 import (
 	"context"
 	"fmt"
+	"io/ioutil"
 	"time"
 
 	"go.uber.org/zap"
 )
 
-type ServiceDiscovery struct {
-	logger *zap.Logger
-	cfg    Config
+type serviceDiscovery struct {
+	logger   *zap.Logger
+	cfg      Config
+	fetcher  *taskFetcher
+	filter   *taskFilter
+	exporter *taskExporter
 }
 
-type ServiceDiscoveryOptions struct {
-	Logger *zap.Logger
+type serviceDiscoveryOptions struct {
+	Logger          *zap.Logger
+	FetcherOverride *taskFetcher // for test
 }
 
-func NewDiscovery(cfg Config, opts ServiceDiscoveryOptions) (*ServiceDiscovery, error) {
-	// NOTE: there are other init logic, currently removed to reduce pr size
-	return &ServiceDiscovery{
-		logger: opts.Logger,
-		cfg:    cfg,
+func newDiscovery(cfg Config, opts serviceDiscoveryOptions) (*serviceDiscovery, error) {
+	svcNameFilter, err := serviceConfigsToFilter(cfg.Services)
+	if err != nil {
+		return nil, fmt.Errorf("init serivce name filter failed: %w", err)
+	}
+	var fetcher *taskFetcher
+	if opts.FetcherOverride != nil {
+		fetcher = opts.FetcherOverride
+	} else {
+		fetcher, err = newTaskFetcher(taskFetcherOptions{
+			Logger:            opts.Logger,
+			Region:            cfg.ClusterRegion,
+			Cluster:           cfg.ClusterName,
+			serviceNameFilter: svcNameFilter,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init fetcher failed: %w", err)
+		}
+	}
+	matchers, err := newMatchers(cfg, MatcherOptions{Logger: opts.Logger})
+	if err != nil {
+		return nil, fmt.Errorf("init matchers failed: %w", err)
+	}
+	filter := newTaskFilter(opts.Logger, matchers)
+	exporter := newTaskExporter(opts.Logger, cfg.ClusterName)
+	return &serviceDiscovery{
+		logger:   opts.Logger,
+		cfg:      cfg,
+		fetcher:  fetcher,
+		filter:   filter,
+		exporter: exporter,
 	}, nil
 }
 
-// RunAndWriteFile writes the output to Config.ResultFile.
-func (s *ServiceDiscovery) RunAndWriteFile(ctx context.Context) error {
+// runAndWriteFile writes the output to Config.ResultFile.
+func (s *serviceDiscovery) runAndWriteFile(ctx context.Context) error {
 	ticker := time.NewTicker(s.cfg.RefreshInterval)
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
 		case <-ticker.C:
-			// do actual work
+			targets, err := s.discover(ctx)
+			if err != nil {
+				// Stop on critical error
+				if cerr := hasCriticalError(s.logger, err); cerr != nil {
+					return cerr
+				}
+				// Print all the minor errors for debugging, e.g. user config etc.
+				printErrors(s.logger, err)
+			}
+			// We may get 0 targets form some recoverable errors
+			// e.g. throttled, in that case we keep existing exported file.
+			if len(targets) == 0 {
```

Comment:
do you think it's better to add a log entry here?

Reply:
We can add a debug/warning log here, most errors are already printed in the error handling above. Though it's possible user wrote a config that does not match any tasks.

Reply:
Updated in https://github.com/pingleig/opentelemetry-collector-contrib/commit/75c4a9c2becb8baf93ae2b03c94aed944585e56b